blocks/blocks.js

```diff
@@ -359,18 +368,242 @@ class Blocks extends Newsletter {
           const { x, y } = script.position
-          possibleDropTarget.acceptDrop(
+          undoEntry.b = possibleDropTarget.acceptDrop(
             script,
             x,
             y,
             snapTo,
-            wrappingC
+            wrappingC,
+            undoEntry
           )
+        }
+        if (!undoEntry.b) {
+          undoEntry.b = script.toJSON().blocks
+        }
+        const extraSteps = this.shoveTarget(undoEntry.b, script)
+        this.addUndoEntry(extraSteps ? [...extraSteps, undoEntry] : undoEntry)
+      }
+    }
+  }
+
+  /**
+   * This gets the target from the given data, displacing it from its original
+   * home if necessary.
+   * @param {number} blockCount - Maximum number of blocks to take (for
+   *   taking from within a stack - required)
+   * @returns {Script}
+   */
+  grabTarget (data, blockCount = 0) {
+    if (Array.isArray(data)) {
+      // Is an array of block JSONs (implying it is a concept to be realized)
+      const script = this.scriptFromJSON({ x: 0, y: 0, blocks: data })
+      script.resize()
+      return script
+    } else if (data.indices) {
+      // Probably means the block was dragged out from a script such that the
+      // old script still remains.
+      const [workspace, scriptIndex, ...indices] = data.indices
+      let component = workspace.scripts[scriptIndex]
+      for (const index of indices) {
+        if (component instanceof Input) {
+          component = component.getValue()
+          if (!(component instanceof Block)) {
+            throw new Error('hwat. The indices point to an input that does not hold a block.')
+          }
+        }
+        if (typeof index === 'number') {
+          component = component.components[index]
+        } else {
+          component = component.getParamComponent(index)
+        }
+      }
+      const script = this.createScript()
+      if (component instanceof Input) {
+        const block = component.getValue()
+        if (!(block instanceof Block)) {
+          throw new Error('hwat. Indices point to an input that does not hold a block.')
+        }
+        component.insertBlock(null)
+        component.resize()
+        script.add(block)
+        return script
+      } else {
+        // `component` now is the block that is clicked on to drag out, so
+        // it and its younger siblings will be deported. Not all though if this
+        // is a reverse of block stack insertion.
+        const parent = component.parent
+        const { dx = 0, dy = 0 } = data
+        const { x, y } = parent.position
+        // Subtraction because this is actually the inverse of the normal operation
+        // which is done in the shove step.
+        parent.setPosition(x - dx, y - dy)
+        let index = indices[indices.length - 1]
+        if (data.branchAround) {
+          const branch = component.getParamComponent(data.branchAround)
+          while (branch.components[0]) {
```

Comment:
I think the while loops deserve some comments quickly explaining what they do, like
```suggestion
          // Add branch contents into script (above the branch block)
          while (branch.components[0]) {
```

Reply:
Epic diff

Reply:
Done!